### ISSUE TITLE
Allow submit any MySQL status variable as Collectd type

### ISSUE THREAD

**Issue by rpv-tomsk** [collectd/collectd#1722]
What is about idea adding option to allow submit _any_  MySQL status variable as Collectd type?
(Possible with addition of predefined/preconfigured 'groups').

Proposed new options for mysql plugin:

`StatusVariable VARIABLE TYPE [TYPE_INSTANCE]`
`StatusVariablePrefix PREFIX TYPE`

Configuration examples, how that can be used:

1)
`StatusVariable Ssl_sessions_reused mysql_ssl_sessions_reused`

This will submit MySQL status variable `Ssl_sessions_reused` as Collectd type `mysql_ssl_sessions_reused`.

2)
`StatusVariablePrefix Table_open_cache_ mysql_table_open_cache`

This will take following variables:
```
mysql> show global status like 'Table_open_cache%';
+----------------------------+------------+
| Variable_name              | Value      |
+----------------------------+------------+
| Table_open_cache_hits      | 1818456442 |
| Table_open_cache_misses    | 989758     |
| Table_open_cache_overflows | 934412     |
+----------------------------+------------+
3 rows in set (0.00 sec)
```
and submit them as type `mysql_table_open_cache` with appropriate type instances `hits`, `misses`, `overflows`.

But, for example, `Ssl_session_cache_*` variables can't be submitted in this way (via StatusVariablePrefix), because of different variables (types) will be mixed together:

```
+-----------------------------+-------+
| Variable_name               | Value |
+-----------------------------+-------+
| Ssl_session_cache_hits      | 0     |
| Ssl_session_cache_misses    | 0     |
| Ssl_session_cache_mode      | NONE  |
| Ssl_session_cache_overflows | 0     |
| Ssl_session_cache_size      | 0     |
| Ssl_session_cache_timeouts  | 0     |
+-----------------------------+-------+
```

Them needs to be configured via explicit `StatusVariable` options with type and type instance set, or we can provide special predefined 'group' for this later.

**** PART 2 **** (unfinished idea)

Syntax: 

`StatusGroup PREDEFINED_GROUP`  (for metrics from 'SHOW GLOBAL STATUS')
`InnoDBStats PREDEFINED_GROUP`  (for InnoDB metrics from information_schema.innodb_metrics)

Benefits of `StatusGroup` over `StatusVariablePrefix` are:

1) Predefined configuration. 
1.1) Different metrics/types can be enabled/submitted in one group.
1.2) If new unsuitable status variables with same prefix will appear in MySQL then no configuration changes will be required, only Collectd upgrade (backward compatability/stable configuration).
2) Submitted types and values can be preprocessed by plugin code to get appropriate values;

As of current `mysql` plugin state, where is following 'groups' exists (Option `InnodbStats` needs to be reviewed separately).

`StatusGroup Statements` for all `Com_*` variables (with `Com_stmt_*` excluded). Zero values excluded.
`StatusGroup Handler` for all `Handler_` variables.  Zero values excluded.
`StatusGroup Qcache` for some of `Qcache_*` variables.
`StatusGroup Bytes` for `Bytes_*` variables.
`StatusGroup Threads` for `Threads_*` variables.
`StatusGroup TableLocks` for all `Table_locks_*` variables.
`StatusGroup Select` for all `Select_*` variables.
`StatusGroup Sort` for all `Sort_*` variables (`mysql_sort_merge_passes`, `mysql_sort_rows` and `mysql_sort` Collectd types).

Some new `groups` which can be added:

`StatusGroup TableOpenCache` for all `Table_open_cache_*` variables as proposed in #1694.
`StatusGroup SSLSessionCache` for Collectd types `mysql_ssl_session_cache` (with type instances `hits/misses/overflow/timeouts`), `mysql_ssl_session_cache` (with type instances `size/used`)

TODO:

1) Work on syntax, decide how to disable groups, enabled by default.
2) Examine InnoDB variables from SHOW GLOBAL STATUS  (fit them into `StatusGroup` schema).
3) Examine InnoDB variables from information_schema.innodb_metrics (function mysql_read_innodb_stats()) and work on similar `InnoDBStats` option.

**Comment by rpv-tomsk** [comment]
Additionally we can recover #431.

@rubenk , @mfournier - what do you think about this PR?

**Comment by octo** [comment]
Hi @rpv-tomsk, thank you very much for opening this! In the future it would be even greater if you could create single-topic issues, i.e. this should have been two issues: one for the generic MySQL status variables and one for the grouping.

I really like the grouping idea because the number of metrics provided by the *MySQL plugin* has grown to a quite large number. Care must be taken to chose a default config, that does not break backwards compatibility. That should not be too hard though.

I'm not convinced we should implement collection of arbitrary metrics: the *DBI plugin* can already do this. Also, the available metrics depend on the SQL statement used (e.g. `SHOW GLOBAL STATUS`, `SHOW MASTER STATUS`), so the config might get quite tricky to set up correctly. That is always a maintenance burden due to user questions.

Best regards,
—octo

**Comment by rpv-tomsk** [comment]
>the DBI plugin can already do this.
At practice, `libdbi` is dead. Yes, it still working. But...
Example: https://sourceforge.net/p/libdbi-drivers/bugs/25/ (unreleased for 5 years).
And I'm using libdbi with sqlite3, yes.

>I'm not convinced we should implement collection of arbitrary metrics: the DBI plugin can already do this. 

It will be better to do all the work in one plugin.

**Comment by rpv-tomsk** [comment]
>1) Work on syntax, decide how to disable groups, enabled by default.

Will that be good, if by default, when no `StatusGroup` options configured, default combination will apply; otherwise explicit configuration is required? I.e, if you want to add one new `StatusGroup`, you must to configure all the rest too.

**Comment by octo** [comment]
> It will be better to do all the work in one plugin.

I disagree with that: each plugin should do *one thing* and do it well. If you need an equivalent to the *DBI plugin* without *libdbi*, that should be its own plugin and not be crammed into the existing *MySQL plugin*.

With regard to the grouping syntax, there are other settings in the *MySQL plugin* that control which metrics are being collected:

* MasterStats
* SlaveStats
* WsrepStats

I'd follow this style for consistency, even though the majority of other plugins uses the *ReportFoobar* schema.

**Comment by rpv-tomsk** [comment]
>I disagree with that: each plugin should do one thing and do it well. 

Yes, 'one thing' is _Mysql status monitoring_, and 'do it well' - allow to configure status reporting which is not supported at the moment.

That is not eqiuvalent to the DBI pugin/libdbi. 
Nobody plans to allow to run _any_ SQL queries in _this_ plugin.

**Comment by rpv-tomsk** [comment]
>I'd follow this style for consistency, even though the majority of other plugins uses the ReportFoobar schema.

Ok, what is your version of naming?

**Comment by octo** [comment]
Sorry @rpv-tomsk, I don't understand what you're asking.

**Comment by rpv-tomsk** [comment]
I have offered:

```
ReportStatus PREDEFINED_GROUP (for metrics from 'SHOW GLOBAL STATUS')
ReportInnoDB PREDEFINED_GROUP (for InnoDB metrics from information_schema.innodb_metrics)
```

As I understand you, you don't like this naming schema/option names.

I want to get your proporsals for naming.

**Comment by octo** [comment]
I don't like that this introduces state, i.e. we will end up with bugs and misunderstandings like #1241. What we need is an easier way to disable a group which is enabled by default.

Most plugins use boolean options for this. For example the *CPU plugin*:

    ReportNumCPU false|true

So, generally plugins use:

    Report<PREDEFINED_GROUP> true|false

The *MySQL plugin* is special in that is already has options like that which don't follow the naming schema:

    WsrepStats  true|false
    MasterStats true|false
    SlaveStats  true|false

Following that, my proposal is:

    <PREDEFINED_GROUP>Stats true|false

**Comment by rpv-tomsk** [comment]
And what is about InnoDB?

```
<PREDEFINED_GROUP>Stats true|false
InnoDB_<PREDEFINED_GROUP>Stats true|false
```

Is that will be ok?

Example:
```
StatementsStats true|false
HandlerStats true|false
QcacheStats true|false
BytesStats true|false
ThreadsStats true|false
TableLocksStats true|false
SelectStats true|false
SortStats true|false
TableOpenCacheStats true|false
SSLSessionCacheStats true|false

InnoDB_IOStats true|false
InnoDB_BufferPoolStats true|false
InnoDB_DMLStats true|false
```
